domains/client/block-builder/src/custom_api.rs

```diff
@@ -0,0 +1,326 @@
+use sp_inherents::InherentData;
+use sp_runtime::traits::{Block as BlockT, HashingFor, NumberFor};
+use sp_runtime::{ApplyExtrinsicResult, ExtrinsicInclusionMode, TransactionOutcome};
+use sp_state_machine::backend::AsTrieBackend;
+use sp_state_machine::{
+    BackendTransaction, DBValue, OverlayedChanges, StateMachine, StorageChanges, TrieBackend,
+    TrieBackendBuilder, TrieBackendStorage,
+};
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+type TrieBackendStorageFor<State, Block> =
+    <State as StateBackend<HashingFor<Block>>>::TrieBackendStorage;
+
+pub(crate) type TrieDeltaBackendFor<'a, State, Block> = TrieBackend<
+    DeltaBackend<'a, TrieBackendStorageFor<State, Block>, HashingFor<Block>>,
+    HashingFor<Block>,
+>;
+
+/// Storage changes are the collected throughout the execution.
+pub struct CollectedStorageChanges<H: Hasher> {
+    /// Storage changes that are captured during the execution.
+    /// Can be used for block import.
+    pub storage_changes: StorageChanges<H>,
+    /// Intermediate storage roots in between the execution.
+    pub intermediate_roots: Vec<H::Out>,
+}
+
+/// Create a new trie backend with memory DB delta changes.
+///
+/// This can be used to verify any extrinsic-specific execution on the combined state of `backend`
+/// and `delta`.
+pub fn create_delta_backend<'a, S, H>(
+    backend: &'a TrieBackend<S, H>,
+    delta: BackendTransaction<H>,
+    post_delta_root: H::Out,
+) -> TrieBackend<DeltaBackend<'a, S, H>, H>
+where
+    S: 'a + TrieBackendStorage<H>,
+    H: 'a + Hasher,
+    H::Out: Codec,
+{
+    create_delta_backend_with_maybe_delta(backend, Some(delta), post_delta_root)
+}
+
+fn create_delta_backend_with_maybe_delta<'a, S, H>(
+    backend: &'a TrieBackend<S, H>,
+    maybe_delta: Option<BackendTransaction<H>>,
+    post_delta_root: H::Out,
+) -> TrieBackend<DeltaBackend<'a, S, H>, H>
+where
+    S: 'a + TrieBackendStorage<H>,
+    H: 'a + Hasher,
+    H::Out: Codec,
+{
+    let essence = backend.essence();
+    let delta_backend = DeltaBackend {
+        backend: essence.backend_storage(),
+        delta: maybe_delta,
+        _phantom: PhantomData::<H>,
+    };
+    TrieBackendBuilder::new(delta_backend, post_delta_root).build()
+}
+
+/// DeltaBackend provides the TrieBackend using main backend and some delta changes
+/// that are not part of the main backend.
+pub struct DeltaBackend<'a, S, H>
+where
+    S: 'a + TrieBackendStorage<H>,
+    H: 'a + Hasher,
+{
+    backend: &'a S,
+    delta: Option<BackendTransaction<H>>,
+    _phantom: PhantomData<H>,
+}
+
+impl<'a, S, H> TrieBackendStorage<H> for DeltaBackend<'a, S, H>
+where
+    S: 'a + TrieBackendStorage<H>,
+    H: 'a + Hasher,
+{
+    fn get(&self, key: &H::Out, prefix: Prefix) -> Result<Option<DBValue>, String> {
+        if let Some(db) = &self.delta
+            && let Some(v) = HashDB::get(db, key, prefix)
+        {
+            Ok(Some(v))
+        } else {
+            Ok(self.backend.get(key, prefix)?)
+        }
```

Comment:
Are we missing a “tombstone” check here?

If a key is present in the backend, but deleted in the overlay, this method returns `Some(old)` - the old value from the backend. But the correct value would be `None`, because the key has been deleted in the overlay, and will be deleted in the backend once consolidated.

(If we never delete any keys, or we always “delete” by overwriting with another value, then there’s no bug here. But I just wanted to check.)

Reply:
Good question.

We are not missing any tombstone checks here. If the key is deleted, it will have a default value.

Reply:
Yes. If the key is written(new/overwrite/delete) during the execution, then the memory DB will have the key present. If the value was not written to, then overlay memory db will not have it and we fallback to supporting backend

As a quick note: this is not new code and has been in the in the codebase for quite sometime powering our invalid state transaition Fraud proof :)

Reply:
> I couldn’t find it documented clearly in the TrieBackendStorage API reference, but they do say to use contains() (and not get()) to check for existence:

Ideally does not matter since get returns an optional value.

we do have implicit tests for this through XDM and pallet domains(not sure yet.)
I did test locally sometime back creating a pallet specifically meant to check the overlay deletion and addition. These never made it. You can actually test this by creating a few storage types and update and check them during the block hooks, initialize and finalize.
This will clearly shows if the deletion works